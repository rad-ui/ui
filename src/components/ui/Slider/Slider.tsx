'use client';

import React, { forwardRef, ElementRef, ComponentPropsWithoutRef, ForwardRefExoticComponent, RefAttributes } from 'react';
import SliderRoot from './fragments/SliderRoot';
import SliderTrack from './fragments/SliderTrack';
import SliderThumb from './fragments/SliderThumb';
import SliderRange from './fragments/SliderRange';

export type SliderElement = ElementRef<'div'>;
export type SliderProps = ComponentPropsWithoutRef<'div'>;

type SliderComponent = ForwardRefExoticComponent<SliderProps & RefAttributes<SliderElement>> & {
    Root: typeof SliderRoot;
    Track: typeof SliderTrack;
    Range: typeof SliderRange;
    Thumb: typeof SliderThumb;
};

const Slider = forwardRef<SliderElement, SliderProps>((_props, _ref) => {
    console.warn('Direct usage of Slider is not supported. Please use Slider.Root, Slider.Track, etc. instead.');
    return null;
}) as SliderComponent;

Slider.displayName = 'Slider';

Slider.Root = SliderRoot;
Slider.Track = SliderTrack;
Slider.Range = SliderRange;
Slider.Thumb = SliderThumb;

export default Slider;
