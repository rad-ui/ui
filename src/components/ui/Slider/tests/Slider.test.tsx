import React from 'react';
import { render, screen } from '@testing-library/react';
import Slider from '../Slider';

describe('Slider Component', () => {
    test('Slider.Root forwards ref', () => {
        const ref = React.createRef<HTMLDivElement>();
        render(
            <Slider.Root ref={ref}>
                <Slider.Track>
                    <Slider.Range>
                        <Slider.Thumb />
                    </Slider.Range>
                </Slider.Track>
            </Slider.Root>
        );
        expect(ref.current).toBeInstanceOf(HTMLDivElement);
    });

    test('Slider.Track forwards ref', () => {
        const ref = React.createRef<HTMLDivElement>();
        render(
            <Slider.Root>
                <Slider.Track ref={ref}>
                    <Slider.Range>
                        <Slider.Thumb />
                    </Slider.Range>
                </Slider.Track>
            </Slider.Root>
        );
        expect(ref.current).toBeInstanceOf(HTMLDivElement);
    });

    test('Slider.Range forwards ref', () => {
        const ref = React.createRef<HTMLDivElement>();
        render(
            <Slider.Root>
                <Slider.Track>
                    <Slider.Range ref={ref}>
                        <Slider.Thumb />
                    </Slider.Range>
                </Slider.Track>
            </Slider.Root>
        );
        expect(ref.current).toBeInstanceOf(HTMLDivElement);
    });

    test('Slider.Thumb forwards ref', () => {
        const ref = React.createRef<HTMLDivElement>();
        render(
            <Slider.Root>
                <Slider.Track>
                    <Slider.Range>
                        <Slider.Thumb ref={ref} />
                    </Slider.Range>
                </Slider.Track>
            </Slider.Root>
        );
        expect(ref.current).toBeInstanceOf(HTMLDivElement);
    });

    test('hidden input is present for accessibility', () => {
        render(
            <Slider.Root defaultValue={50}>
                <Slider.Track>
                    <Slider.Range>
                        <Slider.Thumb />
                    </Slider.Range>
                </Slider.Track>
            </Slider.Root>
        );
        const hiddenInput = screen.getByRole('textbox', { hidden: true });
        expect(hiddenInput).toHaveValue('50');
    });

    test('renders without warnings', () => {
        const warnSpy = jest.spyOn(console, 'warn').mockImplementation(() => {});
        render(
            <Slider.Root>
                <Slider.Track>
                    <Slider.Range>
                        <Slider.Thumb />
                    </Slider.Range>
                </Slider.Track>
            </Slider.Root>
        );
        expect(warnSpy).not.toHaveBeenCalled();
        warnSpy.mockRestore();
    });

    test('matches snapshot', () => {
        const { asFragment } = render(
            <Slider.Root defaultValue={25}>
                <Slider.Track>
                    <Slider.Range>
                        <Slider.Thumb />
                    </Slider.Range>
                </Slider.Track>
            </Slider.Root>
        );
        expect(asFragment()).toMatchSnapshot();
    });
});
