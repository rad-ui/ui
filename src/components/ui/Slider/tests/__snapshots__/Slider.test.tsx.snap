// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="rad-ui-slider"
  >
    <div
      class="rad-ui-slider-track"
    >
      <div
        class="rad-ui-slider-range"
        style="left: calc(25% - 16px); width: calc(25%);"
      >
        <div
          class="rad-ui-slider-thumb"
          style="left: calc(25% - 16px);"
        >
          <span
            class="rad-ui-slider-thumb-value"
          />
          <input
            style="display: none;"
            value="25"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
